### PR TITLE
many: reorg cmd/snapinfo.go into snap and new client/clientutil

### PR DESCRIPTION
**Pull request by pedronis**
To do the latter and disentangle the code from needing systemd all the time, this introduces o/servicestate/StatusDecorator.

See single commits for details.

### CODE REVIEW

File: vendor/vendor.json [hunk 1]
```diff
@@ -110,7 +110,7 @@
 			"revisionTime": "2017-09-28T14:21:59Z"
 		},
 		{
-			"checksumSHA1": "fqejS2llZXw3gLnOYhg7pcSlY+Q=",
+			"checksumSHA1": "Yfj2ZrgRrklXrshCM1RxH5pvUY8=",
```
Comment (stolowski) [inline vendor/vendor.json:113]: Why this change?

File: client/clientutil/snapinfo.go [hunk 2]
```diff
@@ -142,55 +133,13 @@ func ClientAppInfosFromSnapAppInfos(apps []*snap.AppInfo) ([]client.AppInfo, err
...
-					Type:    "timer",
-				})
-			case ".socket":
-				appInfo.Activators = append(appInfo.Activators, client.AppActivator{
-					Name:    sockSvcFileToName[st.UnitName],
-					Enabled: st.Enabled,
-					Active:  st.Active,
-					Type:    "socket",
-				})
-			}
```
Comment (stolowski) [inline client/clientutil/snapinfo.go:193]: Oh... I remember you mentioned something about this a few weeks ago /o\

File: client/clientutil/snapinfo_test.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+func (sd *testStatusDecorator) DecorateWithStatus(appInfo *client.AppInfo, app *snap.AppInfo) error {
+	if appInfo.Snap != app.Snap.InstanceName() || appInfo.Name != app.Name {
+		panic("mismatched")
+	}
+	appInfo.Enabled = true
+	appInfo.Active = true
+	return nil
+}
+
+func (*cmdSuite) TestClientSnapFromSnapInfoApps(c *C) {
```
Comment (stolowski) [inline client/clientutil/snapinfo_test.go:166]: Perhaps it should be renamed to `TestClientSnapFromSnapInfoAppsInactive'?

File: overlord/servicestate/servicestate.go [hunk 4]
```diff
@@ -119,3 +122,76 @@ func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction, conte
...
+}) *StatusDecorator {
+	return &StatusDecorator{
+		sysd: systemd.New(dirs.GlobalRootDir, systemd.SystemMode, rep),
+	}
+}
+
+// DecorateWithStatus adds service status information to the given client.AppInfo associated with the given snap.AppInfo.
+func (sd *StatusDecorator) DecorateWithStatus(appInfo *client.AppInfo, snapApp *snap.AppInfo) error {
+	if appInfo.Snap != snapApp.Snap.InstanceName() || appInfo.Name != snapApp.Name {
+		return fmt.Errorf("misassociated app info %v and client app info %s.%s", snapApp, appInfo.Snap, appInfo.Name)
```
Comment (stolowski) [inline overlord/servicestate/servicestate.go:143]: "internal error: ..." ?

File: overlord/servicestate/servicestate.go [hunk 5]
```diff
@@ -119,3 +122,76 @@ func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction, conte
...
+	}
+	if !snapApp.IsService() {
+		// nothing to do
+		return nil
+	}
+
+	// collect all services for a single call to systemctl
+	serviceNames := make([]string, 0, 1)
```
Comment (stolowski) [inline overlord/servicestate/servicestate.go:151]: Not a problem by any means, but wondering why the change from the original code:
`serviceNames := make([]string, 0, 1+len(app.Sockets)+1)` ?

File: overlord/servicestate/servicestate_test.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+		Name: "app",
+	}
+	snapApp := &snap.AppInfo{Snap: snp, Name: "app"}
+
+	err := sd.DecorateWithStatus(app, snapApp)
+	c.Assert(err, IsNil)
+
+	// service only
+	app = &client.AppInfo{
+		Snap:   "foo",
```
Comment (stolowski) [inline overlord/servicestate/servicestate_test.go:74]: Nitpick, `Snap: snp.SuggestedName`?

File: overlord/servicestate/servicestate_test.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+	err = sd.DecorateWithStatus(app, snapApp)
+	c.Assert(err, IsNil)
+	c.Check(app.Active, Equals, true)
+	c.Check(app.Enabled, Equals, true)
+	c.Check(app.Activators, DeepEquals, []client.AppActivator{
+		{Name: "socket1", Type: "socket", Active: true, Enabled: true},
+	})
+
+}
```
Comment (stolowski) [inline overlord/servicestate/servicestate_test.go:142]: I think this should also test ActiveState=inactive / UnitFileState=disabled combo, or am I missing something?

Comment (pedronis) [reply overlord/servicestate/servicestate_test.go:142]: I can add that,  but those are also the default values

File: overlord/servicestate/servicestate.go (hunk 5 above)
Comment (pedronis) [reply overlord/servicestate/servicestate.go:151]: no, you are right, this got lost because I re-enabled part of this code piece by piece

Comment (pedronis) [reply overlord/servicestate/servicestate.go:151]: did something along those lines

File: overlord/servicestate/servicestate_test.go (hunk 7 above)
Comment (pedronis) [reply overlord/servicestate/servicestate_test.go:142]: covered both enabled = true|false cases

File: overlord/servicestate/servicestate_test.go (hunk 6 above)
Comment (pedronis) [reply overlord/servicestate/servicestate_test.go:74]: used InstanceName

File: overlord/servicestate/servicestate_test.go (hunk 7 above)
Comment (stolowski) [reply overlord/servicestate/servicestate_test.go:152]: Thank you

File: client/clientutil/snapinfo.go [hunk 8]
```diff
@@ -142,55 +133,13 @@ func ClientAppInfosFromSnapAppInfos(apps []*snap.AppInfo) ([]client.AppInfo, err
 		}
 
 		appInfo.Daemon = app.Daemon
-		if !app.IsService() || !app.Snap.IsActive() {
+		if !app.IsService() || decorator == nil || !app.Snap.IsActive() {
```
Comment (mvo5) [inline client/clientutil/snapinfo.go:136]: It was a bit unexpected to me that this decorator is only run for active snaps. It makes sense of course but I wonder if the decision for ignoring inactive snaps should happen inside the decorator? I.e. I wonder if clientutil should know about active/inactive (or even service or not) or if that's knowledge that should be art of the decorator. This way the unit test below would also be slightly easy to read as the decorator code would run for active/non-active in the same way.

File: client/clientutil/snapinfo_test.go [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+	c.Check(ci.Summary, Equals, "the-summary")
+	c.Check(ci.Description, Equals, "the-description")
+	c.Check(ci.Icon, Equals, si.Media.IconURL())
+	c.Check(ci.Website, Equals, si.Website)
+	c.Check(ci.StoreURL, Equals, si.StoreURL)
+	c.Check(ci.Developer, Equals, "thingyinc")
+	c.Check(ci.Publisher, DeepEquals, &si.Publisher)
+}
+
+type testStatusDecorator struct{}
```
Comment (mvo5) [inline client/clientutil/snapinfo_test.go:155]: Should we add something like:
```
-type testStatusDecorator struct{}
+type testStatusDecorator struct {
+       calls int
+}
```
and do "calls++" in testStatusDecorator.DecorateWithStatus so that this can be checked in the tests and we are explicit about it?

File: client/clientutil/snapinfo_test.go [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+	// sanity
+	c.Check(si.IsActive(), Equals, false)
+	// desktop file
+	df := si.Apps["app"].DesktopFile()
+	err := os.MkdirAll(filepath.Dir(df), 0755)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(df, nil, 0644)
+	c.Assert(err, IsNil)
+
+	ci, err := clientutil.ClientSnapFromSnapInfo(si, &testStatusDecorator{})
```
Comment (mvo5) [inline client/clientutil/snapinfo_test.go:190]: I initially stumbled here and did not remember that the testStatusDecorator is only called for active apps. If we keep this, maybe we can add a small comment below like: "// decorator function not called because app not active" or similar?

File: overlord/servicestate/servicestate.go [hunk 11]
```diff
@@ -119,3 +122,80 @@ func Control(st *state.State, appInfos []*snap.AppInfo, inst *Instruction, conte
...
+
+// DecorateWithStatus adds service status information to the given client.AppInfo associated with the given snap.AppInfo.
+func (sd *StatusDecorator) DecorateWithStatus(appInfo *client.AppInfo, snapApp *snap.AppInfo) error {
+	if appInfo.Snap != snapApp.Snap.InstanceName() || appInfo.Name != snapApp.Name {
+		return fmt.Errorf("internal error: misassociated app info %v and client app info %s.%s", snapApp, appInfo.Snap, appInfo.Name)
+	}
+	if !snapApp.IsService() {
+		// nothing to do
+		return nil
+	}
```
Comment (mvo5) [inline overlord/servicestate/servicestate.go:148]: Should we check: `if !snapApp.IsActive() { return nil }` here too?

File: snap/info.go [hunk 12]
```diff
@@ -1355,3 +1355,18 @@ func SortServices(apps []*AppInfo) (sorted []*AppInfo, err error) {
 	}
 	return sorted, nil
 }
+
+// AppInfoBySnapApp supports sorting the given slice of app infos by
```
Comment (mvo5) [inline snap/info.go:1359]: Thanks for moving this here!

File: client/clientutil/snapinfo.go (hunk 8 above)
Comment (pedronis) [reply client/clientutil/snapinfo.go:136]: I agree it would make the code a bit simpler here but OTOH apps don't have a status, so why call a status decorator on something that doesn't have a status? Maybe the doc comment of the functions should make that clearer.

File: overlord/servicestate/servicestate.go (hunk 11 above)
Comment (pedronis) [reply overlord/servicestate/servicestate.go:148]: yes, it's a bit safer

File: client/clientutil/snapinfo.go (hunk 8 above)
Comment (mvo5) [reply client/clientutil/snapinfo.go:136]: Yeah, maybe a small doc comment. Or renaming to "AppDecorator" and running for all apps (active or not) but either way totally fine :)

File: overlord/servicestate/servicestate.go (hunk 11 above)
Comment (pedronis) [reply overlord/servicestate/servicestate.go:148]: this needed test changes to have IsActive return true.

File: client/clientutil/snapinfo_test.go [hunk 13]
```diff
@@ -200,6 +204,8 @@ func (*cmdSuite) TestClientSnapFromSnapInfoAppsInactive(c *C) {
 		},
 		{Snap: "the-snap_insta", Name: "svc", Daemon: "simple"},
 	})
+	// not called on inactive snaps
```
Comment (mvo5) [inline client/clientutil/snapinfo_test.go:207]: :heart:
